File: frontend/lib/api/reports.ts
/**
 * Typed wrapper over the Reports API (backend/app/api/v1/reports.py).
 *
 * Phase 3.2 — entity-layer endpoints only. AI / streaming / conversation
 * / share / export wrappers land alongside their sub-phases.
 *
 * Auth: pulls the bearer JWT from readAdminSession(). All seeded
 * users (internal_admin, reviewer, client_admin, providers via the
 * /portal/enter flow) carry a JWT under the same key.
 */

import { readAdminSession } from "@/lib/session/admin";
import type {
  ReportAudience,
  ReportStatus,
  ReportVersionOrigin,
} from "@/lib/reports/constants";

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://127.0.0.1:8000";

export class ReportsApiError extends Error {
  status: number;
  constructor(status: number, message: string) {
    super(message);
    this.status = status;
    this.name = "ReportsApiError";
  }
}

async function fetchJson<T>(path: string, init: RequestInit = {}): Promise<T> {
  const session = readAdminSession();
  if (!session?.access_token) {
    throw new ReportsApiError(401, "No active session.");
  }
  const headers = new Headers(init.headers ?? {});
  if (!headers.has("Content-Type") && init.body) {
    headers.set("Content-Type", "application/json");
  }
  headers.set("Authorization", `Bearer ${session.access_token}`);
  const response = await fetch(`${API_BASE_URL}${path}`, { ...init, headers });
  if (!response.ok) {
    const detail = await response.text().catch(() => "");
    throw new ReportsApiError(response.status, detail || response.statusText);
  }
  if (response.status === 204) return undefined as unknown as T;
  return (await response.json()) as T;
}

// ─── Wire shapes ─────────────────────────────────────────────────

/**
 * The canvas tree persisted inside a ReportVersion. Schema is open at
 * the entity-API level; per-block schemas are enforced by the
 * registry (see lib/reports/registry.ts).
 */
export interface ReportContent {
  schema_version: number;
  blocks: ReportBlock[];
  global?: Record<string, unknown>;
}

export interface ReportBlock<TConfig = unknown> {
  id: string;
  type: string;
  config: TConfig;
  data?: unknown;
  ai_summary?: AISummary | null;
  layout?: {
    width?: "full" | "half" | "third";
    collapsed?: boolean;
  };
  locked?: boolean;
}

export interface AISummary {
  text: string;
  model: string;
  prompt_hash: string;
  generated_at: string;
  source_snapshot_id: string;
  citations?: Citation[];
}

export interface Citation {
  block_id?: string;
  label: string;
  href?: string;
}

export interface ReportSummary {
  id: string;
  title: string;
  description: string | null;
  audience: ReportAudience;
  status: ReportStatus;
  organization_id: string;
  client_id: string | null;
  vendor_id: string | null;
  current_version_id: string | null;
  created_by_user_id: string;
  created_at: string;
  updated_at: string;
}

export interface ReportVersionSummary {
  id: string;
  report_id: string;
  version_number: number;
  label: string | null;
  parent_version_id: string | null;
  generated_by: ReportVersionOrigin;
  created_by_user_id: string;
  created_at: string;
}

export interface ReportVersionRead extends ReportVersionSummary {
  content_json: ReportContent;
  plan_json: unknown | null;
  source_snapshot_id: string | null;
  llm_metadata: unknown | null;
}

export interface ReportRead extends ReportSummary {
  current_version: ReportVersionRead | null;
}

export interface ReportList {
  items: ReportSummary[];
  total: number;
}

export interface ReportVersionList {
  items: ReportVersionSummary[];
  total: number;
}

// ─── Endpoints ───────────────────────────────────────────────────

export interface CreateReportInput {
  title: string;
  description?: string | null;
  audience: ReportAudience;
  client_id?: string | null;
  vendor_id?: string | null;
  initial_content_json?: ReportContent | null;
}

export function createReport(
  input: CreateReportInput,
  options: { organizationId?: string } = {},
): Promise<ReportRead> {
  const qs = options.organizationId
    ? `?organization_id=${encodeURIComponent(options.organizationId)}`
    : "";
  return fetchJson<ReportRead>(`/api/v1/reports${qs}`, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export interface ListReportsOptions {
  organizationId?: string;
  status?: ReportStatus;
  limit?: number;
  offset?: number;
}

export function listReports(
  options: ListReportsOptions = {},
): Promise<ReportList> {
  const params = new URLSearchParams();
  if (options.organizationId) params.set("organization_id", options.organizationId);
  if (options.status) params.set("status", options.status);
  if (options.limit !== undefined) params.set("limit", String(options.limit));
  if (options.offset !== undefined) params.set("offset", String(options.offset));
  const qs = params.toString();
  return fetchJson<ReportList>(`/api/v1/reports${qs ? `?${qs}` : ""}`);
}

export function getReport(reportId: string): Promise<ReportRead> {
  return fetchJson<ReportRead>(`/api/v1/reports/${reportId}`);
}

export interface PatchReportInput {
  title?: string;
  description?: string | null;
  audience?: ReportAudience;
  status?: ReportStatus;
  client_id?: string | null;
  vendor_id?: string | null;
}

export function patchReport(
  reportId: string,
  input: PatchReportInput,
): Promise<ReportRead> {
  return fetchJson<ReportRead>(`/api/v1/reports/${reportId}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export interface CreateVersionInput {
  content_json: ReportContent;
  label?: string | null;
  plan_json?: unknown | null;
  generated_by?: ReportVersionOrigin;
  parent_version_id?: string | null;
  source_snapshot_id?: string | null;
  llm_metadata?: unknown | null;
}

export function createVersion(
  reportId: string,
  input: CreateVersionInput,
): Promise<ReportVersionRead> {
  return fetchJson<ReportVersionRead>(`/api/v1/reports/${reportId}/versions`, {
    method: "POST",
    body: JSON.stringify({ ...input, generated_by: input.generated_by ?? "user" }),
  });
}

export function listVersions(reportId: string): Promise<ReportVersionList> {
  return fetchJson<ReportVersionList>(`/api/v1/reports/${reportId}/versions`);
}

export function getVersion(
  reportId: string,
  versionNumber: number,
): Promise<ReportVersionRead> {
  return fetchJson<ReportVersionRead>(
    `/api/v1/reports/${reportId}/versions/${versionNumber}`,
  );
}

// ─── Helpers ─────────────────────────────────────────────────────

export function emptyContent(): ReportContent {
  return { schema_version: 1, blocks: [], global: {} };
}

File: apps/api/app/schemas/reports.py
"""Pydantic schemas for the Reports API.

Phase 3.1 ships the entity-level shapes only. Block-level schemas (the
content_json contents) are deliberately left open — the block registry
in Phase 3.2+ will validate those shape-by-shape against zod / Pydantic
schemas registered per block type. For now ``content_json`` is typed as
``dict`` and accepted as-is.

See docs/REPORTS_ARCHITECTURE.md §4 and §13.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.constants.reports import (
    ConversationRole,
    ExportFormat,
    ExportStatus,
    ReportAudience,
    ReportStatus,
    ReportVersionOrigin,
)

# ─── Reports ─────────────────────────────────────────────────────


class ReportCreate(BaseModel):
    """Create-a-report payload.

    Audience is required. If audience != internal_only the caller must
    supply at least one of client_id / vendor_id — enforced by both the
    DB CHECK constraint and the service layer for nicer error copy.
    """

    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    audience: ReportAudience
    client_id: str | None = None
    vendor_id: str | None = None
    # The initial empty canvas. The service layer creates a v1 version
    # for any new report so the editor has something to load.
    initial_content_json: dict | None = None


class ReportPatch(BaseModel):
    """Partial update for report metadata.

    Audience changes are allowed but trigger a service-layer
    re-validation of scope (the same CHECK rule).
    """

    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    audience: ReportAudience | None = None
    status: ReportStatus | None = None
    client_id: str | None = None
    vendor_id: str | None = None


class ReportSummary(BaseModel):
    """List-view row for a report."""

    id: str
    title: str
    description: str | None
    audience: ReportAudience
    status: ReportStatus
    organization_id: str
    client_id: str | None
    vendor_id: str | None
    current_version_id: str | None
    created_by_user_id: str
    created_at: datetime
    updated_at: datetime


class ReportRead(ReportSummary):
    """Full read response. Includes the latest version inline so the
    editor can mount without a second round-trip.
    """

    current_version: ReportVersionRead | None = None


# ─── Versions ────────────────────────────────────────────────────


class ReportVersionCreate(BaseModel):
    """Manual save payload.

    The caller writes the full content_json. The service layer assigns
    version_number = max(existing) + 1 atomically.
    """

    content_json: dict
    label: str | None = Field(default=None, max_length=120)
    plan_json: dict | None = None
    generated_by: ReportVersionOrigin = ReportVersionOrigin.USER
    parent_version_id: str | None = None
    source_snapshot_id: str | None = None
    llm_metadata: dict | None = None


class ReportVersionSummary(BaseModel):
    """List-view row for a version (no content)."""

    id: str
    report_id: str
    version_number: int
    label: str | None
    parent_version_id: str | None
    generated_by: ReportVersionOrigin
    created_by_user_id: str
    created_at: datetime


class ReportVersionRead(ReportVersionSummary):
    """Full version including content_json + plan_json + llm_metadata."""

    content_json: dict
    plan_json: dict | None
    source_snapshot_id: str | None
    llm_metadata: dict | None


# ─── List wrappers ───────────────────────────────────────────────


class ReportList(BaseModel):
    items: list[ReportSummary]
    total: int


class ReportVersionList(BaseModel):
    items: list[ReportVersionSummary]
    total: int


# ─── Conversations / shares / exports (placeholder schemas) ──────
#
# Tables + models exist for these in Phase 3.1, but the endpoints
# don't. The schemas are scaffolded here so the later sub-phases
# don't have to think about wire shapes — they just fill in the
# endpoint code.


class ConversationTurnRead(BaseModel):
    id: str
    report_id: str
    turn_number: int
    role: ConversationRole
    content_json: dict
    attached_version_id: str | None
    created_at: datetime


class ReportShareRead(BaseModel):
    id: str
    report_id: str
    version_id: str
    audience: ReportAudience
    watermark: str | None
    expires_at: datetime | None
    revoked_at: datetime | None
    created_at: datetime
    last_accessed_at: datetime | None
    access_count: int


class ReportExportRead(BaseModel):
    id: str
    report_id: str
    version_id: str
    format: ExportFormat
    status: ExportStatus
    bytes: int | None
    requested_at: datetime
    ready_at: datetime | None
    error_text: str | None


# ─── Presets (R1.0) ──────────────────────────────────────────────


class ReportPresetSummary(BaseModel):
    """One preset entry returned by GET /api/v1/reports/_presets.

    The frontend renders these as cards in the role-scoped report
    list page. ``recommended_prompt`` is the string the editor
    pre-fills into the AI prompt textarea on first open.
    """

    id: str
    title: str
    description: str
    audience: ReportAudience
    recommended_prompt: str


class ReportPresetList(BaseModel):
    items: list[ReportPresetSummary]


class CreateFromPresetRequest(BaseModel):
    """Body for POST /api/v1/reports/from-preset.

    ``organization_id`` is required for callers with more than one
    membership; otherwise we use the single org just like ReportCreate.

    ``client_id`` / ``vendor_id`` are forwarded to the underlying
    create_report call. The service layer rejects audiences other than
    ``internal_only`` unless at least one is supplied. For preset
    callers the endpoint auto-resolves ``client_id`` from the
    caller's client_admin membership when not supplied — see
    ``post_from_preset``.
    """

    preset_id: str = Field(min_length=1, max_length=80)
    client_id: str | None = None
    vendor_id: str | None = None
    # When true, the endpoint generates the report's first populated version
    # inline (hybrid: AI with deterministic fallback) before returning, so the
    # frontend can route straight to a finished read-only report. When false
    # (legacy), it returns an empty report for the editor flow.
    auto_generate: bool = False


# ─── Planner (Phase 3.3a) ────────────────────────────────────────


class PlanReportRequest(BaseModel):
    """Body for ``POST /api/v1/reports/{id}/plan``.

    The user supplies a natural-language description of what they
    want. The server assembles tenant-scoped context and asks the
    LLM to plan the structured block sequence.
    """

    prompt: str = Field(min_length=1, max_length=4000)
    period: str | None = Field(default=None, max_length=20)


class PlannedBlockResponse(BaseModel):
    """One block in a plan response."""

    id: str
    type: str
    config: dict


class PlanReportResponse(BaseModel):
    """``POST /api/v1/reports/{id}/plan`` returns this.

    The plan is not yet executed and no new ``report_versions`` row
    is created — that's Phase 3.3b. The endpoint persists a
    ``compliance_snapshots`` row so the plan can be audited; the
    snapshot_id is included here.
    """

    blocks: list[PlannedBlockResponse]
    rationale: str
    audience: ReportAudience
    scope_hint: dict
    model: str
    stop_reason: str
    usage: dict
    snapshot_id: str
    llm_backend: str
